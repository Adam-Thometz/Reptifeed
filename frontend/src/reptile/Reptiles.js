import React, { useContext } from "react";
import UserContext from "../utils/UserContext";
import { Link, useParams } from "react-router-dom";
import ReptileCard from "./ReptileCard";

const Reptiles = () => {
  const { reptiles } = useContext(UserContext);
  const { id } = useParams();

  document.title = `My Reptiles | Reptifeed`

  return (
    <div className="Reptiles">
      <h1>My Reptiles</h1>
      <p>Click on a reptile to feed them.</p>
      {!reptiles.length ? (
        <h2>You don't have any reptiles. <Link to={`/users/${+id}/reptiles/add`}>Click here to add one</Link></h2>
      ) : reptiles.map(r => (
        <ReptileCard
          key={r.id}
          id={r.id}
          name={r.name}
          image={r.imgUrl}
          species={r.species}
          ownerId={r.ownerId}
          link
        />
      ))}
    </div>
  );
};

export default Reptiles;