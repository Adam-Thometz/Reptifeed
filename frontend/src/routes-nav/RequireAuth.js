import React, { useContext } from "react";
import Unauthorized from "../common/Unauthorized";
import UserContext from "../utils/UserContext";
import { useParams } from 'react-router-dom'

const RequireAuth = ({ children, adminOnly = false }) => {
  const { currUser } = useContext(UserContext);
  const { id } = useParams();

  if (currUser) {
    if (adminOnly) {
      if (currUser.isAdmin) {
        return children;
      } else {
        return <Unauthorized />;
      }
    }
    else {
      if (id) {
        if (currUser.isAdmin || currUser.id === +id) {
          return children;
        } else {
          return <Unauthorized />;
        }
      } else {
        if (currUser) {
          return children;
        } else {
          return <Unauthorized />;
        }
      }
    }
  } else {
    return <Unauthorized />;
  }
}

export default RequireAuth;