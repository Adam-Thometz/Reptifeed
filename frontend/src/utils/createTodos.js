
/** createTodos(reptiles, pantry)
 * 
 * Takes a user's reptiles and pantry and creates a list of things they need to do to unlock the app's full features
 * 
 * Returns an two arrays of Link components, one of important things and one of nice-to-haves
 */

function createTodos(reptiles, pantry) {
  /** Essential checks:
   * - reptileCheck: Checks if a user has reptiles
   * - vegetableOftenCheck: Checks if a user has a vegetable staple in their pantry
   * - proteinOftenCheck: Checks if a user has a protein staple in their pantry
   * - treatModeratelyCheck: Checks if a user has a treat staple in their pantry
   * - supplementCheck: Checks if a user has at least two different supplements
   */
  const essentialTodos = [];

  const reptileCheck = !!reptiles.length;
  const vegetableOftenCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'often');
  const proteinOftenCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'often');
  const treatModeratelyCheck = pantry.some(food => food.isTreat && food.frequency === 'moderately');
  const supplementCheck = (pantry.filter(food => food.type === 'supplement')).length >= 2;

  if (!reptileCheck) essentialTodos.push('Add a reptile');
  if (!vegetableOftenCheck) essentialTodos.push('Add vegetables eaten often');
  if (!proteinOftenCheck) essentialTodos.push('Add protein eaten often');
  if (!treatModeratelyCheck) essentialTodos.push('Add a treat eaten moderately');
  if (!supplementCheck) essentialTodos.push('Add at least two supplements');

  /** Nice-to-have checks:
   * - vegetableModeratelyCheck: Checks if a user has a vegetable that's eaten moderately in their pantry
   * - vegetableOccasionallyCheck: Checks if a user has a vegetable that's eaten occasionally in their pantry
   * - proteinModeratelyCheck: Checks if a user has a protein that's eaten moderately in their pantry
   * - proteinOccasionallyCheck: Checks if a user has a protein that's eaten occasionally in their pantry
   * - treatOccasionallyCheck: Checks if a user has a treat that's eaten rarely in their pantry
   */
  const niceToHaveTodos = [];

  const vegetableModeratelyCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'moderately');
  const vegetableOccasionallyCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'occasionally' && !food.isTreat);
  const proteinModeratelyCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'moderately');
  const proteinOccasionallyCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'occasionally' && !food.isTreat);
  const treatOccasionallyCheck = pantry.some(food => food.isTreat && food.frequency === 'occasionally');

  if (!vegetableModeratelyCheck) niceToHaveTodos.push('Add a vegetable eaten moderately');
  if (!vegetableOccasionallyCheck) niceToHaveTodos.push('Add a vegetable eaten occasionally');
  if (!proteinModeratelyCheck) niceToHaveTodos.push('Add a protein eaten moderately');
  if (!proteinOccasionallyCheck) niceToHaveTodos.push('Add a protein eaten occasionally');
  if (!treatOccasionallyCheck) niceToHaveTodos.push('Add a treat eaten occasionally');

  return {essentialTodos, niceToHaveTodos};
};

export default createTodos;