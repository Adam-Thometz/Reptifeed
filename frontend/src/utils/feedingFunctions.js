/** Feeding functions
 * 
 * Used to get foods from pantry that match frequencies and food type.
*/

function getFoodFromPantry(pantry, freq, type) {
  const availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === freq && food.type === type)));
  if (!availableFood.length) return `Get some more ${type}`
  const selectedIdx = Math.floor(Math.random() * availableFood.length);

  return availableFood[selectedIdx];
};

function getTreatFromPantry(pantry, freq) {
  const availableTreats = pantry.filter(food => food.isTreat && food.frequency === freq);
  if (!availableTreats.length) return "You don't have any treats";
  const selectedIdx = Math.floor(Math.random() * availableTreats.length); 
  
  return availableTreats[selectedIdx];
};

export { getFoodFromPantry, getTreatFromPantry };