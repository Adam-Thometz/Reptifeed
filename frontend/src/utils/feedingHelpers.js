/** Feeding functions
 * 
 * Used to get foods from pantry that match frequencies and food type.
 * 
 * If no foods are available for the frequency, returns foods eaten most often
*/

function getFoodFromPantry(pantry, freq, type) {
  let availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === freq && food.type === type)));
  if (!availableFood.length) {
    availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === 'often' && food.type === type)));
  };
  const selectedIdx = Math.floor(Math.random() * availableFood.length);

  return availableFood[selectedIdx];
};

function getTreatFromPantry(pantry, freq) {
  let availableTreats = pantry.filter(food => food.isTreat && food.frequency === freq);
  if (!availableTreats.length) {
    availableTreats = pantry.filter(food => food.isTreat);
  };
  const selectedIdx = Math.floor(Math.random() * availableTreats.length); 
  
  return availableTreats[selectedIdx];
};

function getNextSupplement(pantry, lastSupplement) {
  const nextSupplement = lastSupplement === 'calcium' ?
    pantry.find(food => food.name === 'multivitamin') :
    pantry.find(food => food.name === 'calcium');
  
  return nextSupplement;
}

export { getFoodFromPantry, getTreatFromPantry, getNextSupplement };