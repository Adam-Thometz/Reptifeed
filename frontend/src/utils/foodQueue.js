// foodQueue: Feed reptile based on frequency
const foodQueue = ["often", "often", "moderately", "often", "often", "moderately", "often", "often", "moderately", "occasionally"];

// treatQueue: Give reptile treats based on frequency
const treatQueue = ["moderately", "moderately", "occasionally"];

export { foodQueue, treatQueue };