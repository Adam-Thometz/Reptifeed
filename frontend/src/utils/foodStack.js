// foodStack: Feed reptile based on frequency
const foodStack = ["occasionally", "moderately", "often", "often", "moderately", "often", "often", "moderately", "often", "often"];

// treatStack: Give reptile treats based on frequency
const treatStack = ["occasionally", "moderately", "moderately"];

export { foodStack, treatStack };