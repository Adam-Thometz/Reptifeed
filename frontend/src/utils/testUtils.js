import React from "react";
import UserContext from "./UserContext";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

/** Test user, reptiles, and pantry/food */

const testUser = {
  id: 1,
  username: 'rickSanchez',
  password: 'ilikeMorty',
  email: 'rickman@qmail.com',
  isAdmin: false
};

const testReptile = [
  {
    id: 1,
    name: 'two crows',
    species: 'crows',
    subspecies: 'two of them',
    birthday: '2015-10-29',
    imgUrl: '',
    ownerId: 1
  }
];

const testPantry = [
  {
    name: 'arugula',
    type: 'vegetable',
    frequency: 'often',
    image: 'https://halfyourplate-4kgxi1gvwldjzby.stackpathdns.com/wp-content/uploads/2014/12/arugula_small.gif',
    isTreat: false,
    tips: ''
  }
];

/** UserContext mock
 * 
 * Wrap components in UserProvider in test files if UserContext is used by a component
 */

const UserProvider = ({ children, currUser = testUser, pantry = testPantry, reptiles = testReptile }) => (
  <UserContext.Provider value={{currUser, pantry, reptiles}}>
    {children}
  </UserContext.Provider>
);

/** custom render
 * 
 * wraps component in MemoryRouter. Useful for all components.
 */

const MemoryRouterWithInitialRoutes = ({ children, initialRoutes }) => {
  return (
    <MemoryRouter initialEntries={initialRoutes}>
      {children}
    </MemoryRouter>
  )
}

const MemoryRouterWithInitialRoutesAndUser = ({ children, initialRoutes }) => {
  return (
    <MemoryRouter initialEntries={initialRoutes}>
      {children}
    </MemoryRouter>
  )
}

const customRender = (component, options) => {
  const initialRoutes = (options && options.initialRoutes)
    ? options.initialRoutes
    : ['/'];
  return render(component, { wrapper: (args) =>
    MemoryRouterWithInitialRoutes({
      ...args,
      initialRoutes
    }), ...options
  });
};

// This export command overrides the render function from the testing library so that we may call our custom render function 'render'
export * from '@testing-library/react'

export { 
  UserProvider,
  customRender as render,
  testUser as user,
  testReptile as reptiles,
  testPantry as pantry
};