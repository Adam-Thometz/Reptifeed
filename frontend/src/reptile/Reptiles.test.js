import React from "react";
import Reptiles from "./Reptiles";
import { render, UserProvider } from "../utils/testUtils";
import RequireAuth from "../routes-nav/RequireAuth";
import { Routes, Route } from 'react-router-dom';

test('it renders without crashing', () => {
  render(<UserProvider>
    <Routes>
      <Route path='/users/:id/reptiles' element={<RequireAuth><Reptiles /></RequireAuth>} />
    </Routes>
  </UserProvider>, { initialRoutes: ['/users/1/reptiles'] });
});

test('it matches snapshot for logged in user', () => {
  const { asFragment } = render(
    <UserProvider>
      <Routes>
        <Route path='/users/:id/reptiles' element={<RequireAuth><Reptiles /></RequireAuth>} />
      </Routes>
    </UserProvider>, { initialRoutes: ['/users/1/reptiles'] }
  );
  expect(asFragment()).toMatchSnapshot();
});

test('it matches snapshot for logged out user', () => {
  const { asFragment } = render(
    <UserProvider currUser={null}>
      <Routes>
        <Route path='/users/:id/reptiles' element={<RequireAuth><Reptiles /></RequireAuth>} />
      </Routes>
    </UserProvider>, { initialRoutes: ['/users/1/reptiles'] }
  );
  expect(asFragment()).toMatchSnapshot();
});