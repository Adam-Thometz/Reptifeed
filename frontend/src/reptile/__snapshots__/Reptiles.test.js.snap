// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot for logged in user 1`] = `
<DocumentFragment>
  <div
    class="Routes"
  >
    <div
      class="Reptiles"
    >
      <h1>
        My Reptiles
      </h1>
      <h4>
        Click on a reptile to feed them.
      </h4>
      <div
        class="ReptileCard"
      >
        <img
          alt="two crows"
          src=""
        />
        <h4>
          two crows
        </h4>
        <h5>
          crows
        </h5>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`it matches snapshot for logged out user 1`] = `
<DocumentFragment>
  <div
    class="Routes"
  >
    <div
      class="Unauthorized"
    >
      <h4>
        You are not authorized to view this page! Here are your options:
      </h4>
      <div
        class="Unauthorized-options"
      >
        <a
          href="/login"
        >
          Log in to your account
        </a>
        <a
          href="/register"
        >
          Register for an account with us
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
